src/lambda_toolshed/papillon.cljc

```diff
@@ -43,146 +47,105 @@
   "Transition the context `ctx` to the candidate context value `candidate`.  This function
   works synchronously with value candidates -any async processing should be performed prior
   to invoking this function."
-  [ctx candidate-ctx tag]
+  [ctx tag candidate-ctx]
   (cond
     (-> candidate-ctx reduced?) (-> ctx
-                                    (transition (unreduced candidate-ctx) tag)
+                                    (transition tag (unreduced candidate-ctx))
                                     clear-queue)
     (-> candidate-ctx error?) (-> ctx
                                   (assoc ::error candidate-ctx)
+                                  (assoc ::stage :error)
                                   clear-queue)
-    (-> candidate-ctx context? not) (-> ctx
-                                        (assoc ::error (ex-info (fmt "Context was lost at %s!" tag)
-                                                                {::tag tag
-                                                                 ::ctx ctx
-                                                                 ::candidate-ctx candidate-ctx}))
-                                        clear-queue)
-    :else candidate-ctx))
-
-(defn- try-stage
-  "Try to invoke the stage function at the `stage` key of the interceptor `ix`
-  with the context `ctx` and return the result.  If there is no value at the
-  `stage` key the context is returned unchanged.
-
-  If the `:lambda-toolshed.papillon/trace` key is present then stage function
-  invocations are conj'd onto its value."
-  [{trace ::trace :as ctx} ix stage]
-  (let [tag [(or (:name ix) (-> ix meta :name)) stage]
-        ctx (if trace
-              (update ctx ::trace conj tag)
-              ctx)]
-    (let [f (or (stage ix) identity)]
-      (try
-        (let [res (f ctx)]
-          (if (satisfies? ReadPort res)
-            (go (transition ctx (<! res) tag))
-            (transition ctx res tag)))
-        (catch #?(:clj Throwable :cljs :default) err
-          (transition ctx err tag))))))
-
-(defn- enter
-  "Run the queued enter chain in the given context `ctx`.  If the
-  `:lambda-toolshed.papillon/error` key is present in `ctx` then clear the queue
-  (thus terminating the `enter` chain) and start processing the `:error` chain.
-  If `ctx` is reduced (per `clojure.core/reduced?`) an early return is inferred
-  and the unreduced `ctx` is used to start processing the `:leave` chain.
-
-  `enter` also 'collapses' nested async calls by recurisvely calling
-  itself with the value taken from the channel to ensure that if a
-  channel was returned, it doesn't halt if that channel is another
-  channel, but continues until it gets a non-ReadPort value for the
-  context."
+    (-> candidate-ctx context?) (let [{::keys [error stage]} candidate-ctx]
```

Comment:
I can take care of it in a separate PR, though the thread first use on these predicate checks for the `cond` don't really seem all that necessary anymore.

Reply:
Agreed.  Originally (some of) these statements were multistep -but no longer.  Fix coming.

Reply:
ok, cool 👍